### PR TITLE
Memoize Addressable::URI instance variables

### PR DESCRIPTION
**Pull request by sferik**
In the case where a fragment is present but empty (e.g. "http://example.com#"), the `normalized_fragment` method returns `nil`, so it can't be memoized with `||=`. This [patch](https://github.com/sferik/addressable/commit/67e36a4c7f406b874281037b3454d2cfd19ddc44) fixes this problem, so the `@normalized_fragment` instance variable is only ever assigned once, even when the method returns `nil`. This was causing an issue with frozen Addressable::URI (see https://github.com/sferik/t/issues/197).

I hope you’ll find a moment to release this fix during your adventures in Tanzania.

Cheers! :beers:

### CODE REVIEW

**Comment by sporkmonger** [comment]
Tests need to pass.

**Comment by sferik** [comment]
@sporkmonger Working on it. It turns out that Ruby’s `defined?` keyword, passed an instance variable, behaves differently from `Object#instance_variable_defined?`. This was very surprising to me.

**Comment by sferik** [comment]
@sporkmonger Never mind. That’s not actually true. I was using `defined?` incorrectly.

**Comment by sferik** [comment]
Not sure what all the @coveralls noise is about but it should be good to merge now.

**Comment by sferik** [comment]
Okay, now coveralls is failing but only because the total number of lines of code went down in a file with 100% c0 coverage. I don’t think that should count against this pull request. Thanks for considering it.

**Comment by sporkmonger** [comment]
Wow that coveralls spam is annoying.

File: spec/addressable/template_spec.rb [hunk 1]
```diff
@@ -1293,6 +1323,12 @@ def self.match(name)
     end
   end
 
+  describe 'inspect' do
+    it "returns a string representation of the MatchData's state" do
+      expect(its.inspect).to match(/\A#<Addressable::Template::MatchData:0x[0-9a-f]{0,12} RESULT:{"foo"=>"ab", "bar"=>"cd"}>\z/)
```
Comment (sporkmonger) [inline spec/addressable/template_spec.rb]: Line is too long. Also not clear why this is related to the rest of the pull request.

**Comment by sporkmonger** [comment]
I gotta tell ya, I appreciate that you put a lot of work into this, but I still can't merge this.

Drop the bundle changes, drop the gemspec and dependency changes, drop essentially everything that's not either a change to the `lib` directory or the `spec` directory, or directly dependent on a change made to one or the other.

Pull requests should always change the least amount of code possible, and they should always change only one thing at a time.

File: spec/addressable/template_spec.rb [hunk 2]
```diff
@@ -728,6 +728,21 @@ def self.match(name)
...
 describe Addressable::Template do
+  describe 'initialize' do
+    context 'with a non-string' do
+      it 'raises a TypeError' do
+        expect { Addressable::Template.new(nil) }.to raise_error(TypeError)
+      end
+    end
+  end
+
+  describe 'freeze' do
```
Comment (sporkmonger) [inline spec/addressable/template_spec.rb]: This is the only template spec that appears to be related to freezing a template. This should be the only template spec in this PR.

File: spec/addressable/template_spec.rb [hunk 3]
```diff
@@ -818,6 +833,21 @@ def self.match(name)
       its(:captures){ should == ["foo", nil, nil] }
     end
   end
+
+  describe 'match' do
```
Comment (sporkmonger) [inline spec/addressable/template_spec.rb]: These are good tests, but they should be in a different PR.

File: spec/addressable/template_spec.rb (hunk 1 above)
Comment (sferik) [inline spec/addressable/template_spec.rb]: This is related to the rest of the pull request insofar as Coveralls was failing because I removed a few lines of covered code, meaning the overall percentage of covered code went down. This covers some uncovered code, which making Coveralls happy again. I’ll remove it or move it into another pull request if you don’t mind merging this pull requests with Coveralls failing.

**Comment by sferik** [comment]
> Pull requests should strive to change the least amount of code possible, and they should change only one thing at a time. This PR is problematic for the latter reason.

I’ve split this into three separate pull requests:

1. This pull request aims to capture only the original intent: **to memoize `Adressable::URI` instance variables**.
2. The additions to the `Addressable::Template` spec, which were only added to make Coveralls :green_heart:, have been moved into https://github.com/sporkmonger/addressable/pull/190.
3. The changes to optimize the CI build are captured in https://github.com/sporkmonger/addressable/pull/189.

I believe these three pull requests new adhere to the principle you’ve stated above but please let me know if you’d like me to subdivide these pull requests any further. Also, feel free to cherry pick individual commits from this pull request (or the others), without merging it, whole :pig2:.

Thank you for taking the time to consider these changes.

**Comment by sporkmonger** [comment]
Amazing thanks so much, I know that was probably a good bit of work.

File: lib/addressable/uri.rb [hunk 4]
```diff
@@ -2356,5 +2363,10 @@ def split_path(path)
       splitted << EMPTY_STR if path.end_with? SLASH
       splitted
     end
+
+    def reset_uri_string_and_hash
```
Comment (sporkmonger) [inline lib/addressable/uri.rb]: Please add YARD documentation for this method.

**Comment by sporkmonger** [comment]
Everything looks good except for the missing YARD documentation.

File: lib/addressable/uri.rb (hunk 4 above)
Comment (sporkmonger) [inline lib/addressable/uri.rb]: Actually, I think I'd prefer a different method name too. The only thing these two have in common is that they're essentially composite values for the entire URI. Method name should probably reflect that rather than being quite so literal. Especially since it's very possible that there would be future additions to the list and in Addressable 3, subclass overrides to this method.

`remove_composite_instance_variables` seems a little too verbose, maybe `remove_composite_values`?

Comment (sferik) [inline lib/addressable/uri.rb]: I’m happy to rename the method and add YARD documentation. I’ll do that now. I was a bit fast and loose with the naming an documentation because it’s a private method.

**Comment by sferik** [comment]
This pull request should be good to merge as soon as the build turns green.

**Comment by sferik** [comment]
Any addition feedback on this one?

**Comment by sporkmonger** [comment]
Nope! Everything looks good!
